librbd: asynchronous image removal state machine

yangdongsheng opened the pull request:
Thanx @vshankar give me the honor to open a separate Pull Request for Async remove state machine. This will be referred by #10896 and #12041

yangdongsheng commented:
Seems it does not compile because there is some other dependent commits. Seems not difficult,  Will address it tomorrow .

yangdongsheng commented:
anyone can help to add tags of rbd and feature? Thanx :)

trociny commented:
@yangdongsheng @vshankar It looks like it needs  some work -- see Jenkins failure.

vshankar commented:
@trociny  mostly yes -- probably due to changes in other parts of librbd.

@yangdongsheng I can take a look at this next week but feel free to fix it!

yangdongsheng commented:
Thanx @trociny 

@vshankar I will try to fix it later after I finish my another work, maybe tomorrow. Thanx

yangdongsheng commented:
@trociny and @vshankar It's green now. I missed one commit about force remove. It's added back now.

trociny commented:
@yangdongsheng @vshankar 
When removing non-existent image, it does not return error:

```
zhuzha:~/ceph/ceph.upstream/build% ./bin/rbd remove NOTEXIST ; echo $?
2016-11-30 11:28:06.677754 7fe8feffd700 -1 librbd::image::RemoveRequest: 0x7fe940170370 handle_open_image error opening image: (2) No such file or directory
Removing image: 100% complete...done.
0

zhuzha:~/ceph/ceph.upstream/build% RBD_FEATURES=127 LD_LIBRARY_PATH=./lib PYTHONPATH=../src/pybind:../src/test/pybind:./lib/cython_modules/lib.2  nosetests -v test_rbd -m 'test_remove_dne'
test_rbd.test_remove_dne ... FAIL

======================================================================
FAIL: test_rbd.test_remove_dne
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/SSD/ceph/ceph.upstream/build/test_rbd.py", line 255, in test_remove_dne
AssertionError: ImageNotFound not raised
```

yangdongsheng commented on src/librbd/image/RemoveRequest.cc:
```diff
@@ -0,0 +1,639 @@
+  ldout(m_cct, 20) << ": r=" << *result << dendl;
+
+  if (*result < 0) {
+    lderr(m_cct) << "error opening image: " << cpp_strerror(*result) << dendl;
```
I mean, the missing of m_retval assignment is definitely a bug. but why the journal and objectmap are not removed in RemoveRequest. @vshankar

vshankar replied:
@yangdongsheng Missing `m_retval` assignment is definitely a bug, nothing intentional. While that needs to be fixed, make sure `m_retval` is not clobbered again in `handle_mirror_image_remove()`.

Journal and ObjectMap are removed -- see `disable_features` which invokes `DisableFeaturesRequest` state machine.

yangdongsheng commented:
@trociny @vshankar The bug about removing NOTEXIST was fixed. And I squashed this fix into the main commit by adding a comment in commit message above my singed-off-by.

trociny commented:
@yangdongsheng @vshankar This test from qa/workunits/rbd/copy.sh fails:
```
test_remove() {
   ...
    # remove with header missing (old format)
    rbd create --image-format 1 -s 1 test1
    rados rm -p rbd test1.rbd
    rbd rm test1
```
Now it returns:
```
2016-12-03 19:19:21.688776 7fec30ff9700 -1 librbd::image::RemoveRequest: 0x7fec6e1b0480 handle_open_image error opening image: (2) No such file or directory
Removing image: 0% complete...failed.
rbd: delete error: (2) No such file or directory
```

yangdongsheng commented:
@trociny  When I look into more about the RemoveRequest, I think there are some more work we can do before merging it. So I propose we hold a while on this pull request. I will send a commit about my change about it for review, if that's okey, I will squash it into the original commit. Sounds good? @trociny @vshankar

yangdongsheng commented:
@trociny  The new commit will address the problem you mentioned. @vshankar please help to review it.

vshankar commented:
> @trociny The new commit will address the problem you mentioned. @vshankar please help to review it.

Will review it soon (currently I'm in mid of pushing my PR). Thanks.

vshankar commented on src/librbd/image/RemoveRequest.cc:
```diff
@@ -0,0 +1,642 @@
+  ldout(m_cct, 20) << ": r=" << *result << dendl;
+
+  if ((*result < 0) && (*result != -ENOENT)) {
+    if (m_retval == 0) {
```
Seems like an unnecessary check. When we are here, `m_retval` is always -ve.

yangdongsheng replied:
Yes, it should be non-zero here. I added a check here just be sure we will not overwrite it. And make this function safe enough even when we move it to another locate in state machine.

That means, yes in this context, we don't need this block of code. But from the insight of making this function logically consistent. what about keep these code here?

yangdongsheng replied:
Yes, it should be non-zero here. I added a check here just be sure we will not overwrite it. And make this function safe enough even when we move it to another locate in state machine.

That means, yes in this context, we don't need this block of code. But from the insight of making this function logically consistent. what about keep these code here?

yangdongsheng replied:
comment sounds good. thanx

trociny commented on src/librbd/image/RemoveRequest.cc:
```diff
@@ -0,0 +1,642 @@
+Context *RemoveRequest<I>::handle_mirror_image_remove(int *result) {
+  ldout(m_cct, 20) << ": r=" << *result << dendl;
+
+  if ((*result < 0) && (*result != -ENOENT)) {
```
Please look at the original code. We need to check for -EOPNOTSUPP too, just in case we are running on an older cluster. Also, I don't like very much unnecessary extra parenthesis.

yangdongsheng commented:
@trociny and @vshankar retest this please, I have addressed these comments and squashed my new commit to the original commit.

yangdongsheng commented:
Hi @trociny , sorry for the late update, I was struggling with some other work. Now this branch is rebased against master, please consider merging it if all looks okey to you. thanx .

trociny commented:
@yangdongsheng I observe the issue with remove not returning ENOENT for not existent image again.

I suggest adding a unittest (test/librbd/test_librbd.cc) that would check this case.

yangdongsheng commented:
@trociny Yes, a unittest sounds very necessary. Will investigate this problem and add a unittest soon.

yangdongsheng commented:
@trociny  the fix for the problem you mentioned is already pushed, and the unittest is still in cooking, will commit later.

yangdongsheng commented:
@trociny I added a test in qa/workunit/rbd/, because the ImageCtx is created in RemoveRequest, I am not sure how to mock it in src/test/librbd/test_librbd.cc. BTW, I renamed the copy.sh in work unit to generic.sh as it is testing more more than rbd copy. Review and retest this please.

trociny commented on qa/workunits/rbd/copy.sh:
```diff
@@ -164,6 +164,7 @@ test_remove() {
     echo "testing remove..."
     remove_images
 
+    echo `rbd remove "NOT_EXIST" > /dev/null 2>&1 || echo $?` | grep 2
```
I would simplify this and check that it return non-exit status (not exactly 2). Something like we have for shrink check:
```
rbd resize testimg1 --size=128 && exit 1 || true   # shrink should fail
```

yangdongsheng replied:
okey thanx

trociny commented on src/librbd/image/RemoveRequest.cc:
```diff
@@ -69,6 +69,8 @@ Context *RemoveRequest<I>::handle_open_image(int *result) {
     lderr(m_cct) << "error opening image: " << cpp_strerror(*result) << dendl;
     m_retval = *result;
+    if (*result != -ENOENT)
+      m_retval = *result;
```
I can't understand this change

yangdongsheng replied:
This is a fix for http://tracker.ceph.com/issues/18315

trociny replied:
I still don't understand your fix. Have you tested this? This patch does not look like have any effect to me, because it still sets m_retval = *result one line above (and still proceed with the removal). And even looking at your #12638 I don't think this is what Jason suggests as a fix for this issue. I had impression he wanted we check for incompatible features, while you just make it to abort when it fails to open the image.

I wouldn't add the patch for this issue here until it is fixed separately.

yangdongsheng replied:
yes, should not add it here. will remove it at first, then I will focus on how to fix it properly in #12638 . thanx for your suggestion.

trociny commented:
> I added a test in qa/workunit/rbd/, because the ImageCtx is created in RemoveRequest, I am not sure how to mock it in src/test/librbd/test_librbd.cc

I

yangdongsheng commented:
@trociny yes, about the test_librbd.cc point, I see what you suggested. But I found the retval of rbd_remove() is always 0.
```
diff --git a/src/test/librbd/test_librbd.cc b/src/test/librbd/test_librbd.cc
index 2e67032..1534e03 100644
--- a/src/test/librbd/test_librbd.cc
+++ b/src/test/librbd/test_librbd.cc
@@ -770,6 +770,7 @@ TEST_F(TestLibRBD, TestCreateLsDelete)
   ASSERT_EQ(0, create_image(ioctx, name2.c_str(), size, &order));
   ASSERT_EQ(2, test_ls(ioctx, 2, name.c_str(), name2.c_str()));
   ASSERT_EQ(0, rbd_remove(ioctx, name.c_str()));
+  ASSERT_EQ(-ENOENT, rbd_remove(ioctx, "NOT_EXIST"));
   ASSERT_EQ(1, test_ls(ioctx, 1, name2.c_str()));
 
   rados_ioctx_destroy(ioctx);


/data/ceph/src/test/librbd/test_librbd.cc:773: Failure
      Expected: -2
To be equal to: rbd_remove(ioctx, "NOT_EXIST")
      Which is: 0
[  FAILED  ] TestLibRBD.TestCreateLsDelete (25 ms)
[----------] 1 test from TestLibRBD (25 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test case ran. (33 ms total)
[  PASSED  ] 0 tests.
[  FAILED  ] 1 test, listed below:
[  FAILED  ] TestLibRBD.TestCreateLsDelete

 1 FAILED TEST
```
So I tried another way to do the test in workunit.

trociny commented:
> ASSERT_EQ(-ENOENT, rbd_remove(ioctx, "NOT_EXIST"));

I would suggest use name.c_str() instead of "NOT_EXIST" -- we just removed this image so we are sure it does not exist. Also this test can be one line below, after test_ls.

The unittest version currently does not work due to an issue with tmap_update from librados test stub implementation -- I have just  created PR for this (#12667). ceph_test_librbd works though.

yangdongsheng commented:
@trociny pushed a new commit about unittest, and I will rebase when your #12667 merged.

trociny commented:
@yangdongsheng  After renaming copy.sh you also need to update qa/suites/rbd/cli/workloads/rbd_cli_copy.yaml and  src/test/run-rbd-tests.

yangdongsheng commented:
@trociny yes, you are right, updated now. thanx.

trociny commented on src/librbd/image/RemoveRequest.cc:
```diff
@@ -623,9 +623,11 @@ template<typename I>
+      lderr(m_cct) << "error removing image from v2 directory: "
+		   << cpp_strerror(*result) << dendl;
+    }
     return m_on_finish;
```
@yangdongsheng After this change the assignment `*result = 0` looks like not needed any more. And we can remove this return statement here then too.

yangdongsheng replied:
done

dillaman commented:
@yangdongsheng Nit: can you rename 'qa/workunits/rbd/generic.sh' to something like 'qa/workunits/rbd/cli_generic.sh' to indicate that it's purpose is to test basic CLI commands?

dillaman commented on src/librbd/CMakeLists.txt:
```diff
@@ -40,6 +40,7 @@ set(librbd_internal_srcs
   image/RefreshParentRequest.cc
   image/RefreshRequest.cc
   image/SetFlagsRequest.cc
+  image/RemoveRequest.cc
```
Nit: try to preserve alphabetical ordering -- move between "RefreshRequest" and "SetFlagsRequest"

yangdongsheng replied:
Done

dillaman commented on src/librbd/image/RemoveRequest.h:
```diff
@@ -0,0 +1,127 @@
+#include "librbd/journal/TypeTraits.h"
+#include "cls/journal/cls_journal_types.h"
+
+using librados::IoCtx;
```
Nit: avoid adding using declarations to header files

yangdongsheng replied:
okey move it to source file.

dillaman commented on src/librbd/image/RemoveRequest.h:
```diff
@@ -0,0 +1,127 @@
+#define CEPH_LIBRBD_IMAGE_REMOVE_REQUEST_H
+
+#include "include/int_types.h"
+#include "include/rbd/librbd.hpp"
```
Double-check these includes -- I think a lot are not required in this header file since I don't see them being used.

yangdongsheng replied:
Done

dillaman commented on src/librbd/image/RemoveRequest.cc:
```diff
@@ -0,0 +1,641 @@
+  m_ioctx.dup(ioctx);
+  m_cct = reinterpret_cast<CephContext *>(m_ioctx.cct());
+
+  m_image_ctx = new I((m_image_id.empty() ? m_image_name : std::string()),
```
I think this is getting leaked on the error paths -- definitely on the open image failure case.

yangdongsheng replied:
This will be deleted in switch_thread_context. and this function is on the error path. it would be more clear in state diagram.

dillaman replied:
Even in ```handle_open_image```? It seems that will proceed to remove the mirror image instead of switching contexts and freeing the pointer.

yangdongsheng replied:
yes, remove_mirror_image() will continue to switch_thread_context().

dillaman commented on src/librbd/image/RemoveRequest.cc:
```diff
@@ -0,0 +1,641 @@
+Context *RemoveRequest<I>::handle_open_image(int *result) {
+  ldout(m_cct, 20) << ": r=" << *result << dendl;
+
+  if (*result < 0) {
```
Need to add new test for "*result != -ENOENT" and abort the removal

yangdongsheng replied:
Yea, only continue on ENOENT.

dillaman commented on src/librbd/image/RemoveRequest.h:
```diff
@@ -0,0 +1,127 @@
+  void send();
+
+private:
+  RemoveRequest(IoCtx &ioctx, const std::string &image_name, const std::string &image_id,
```
Please add a state diagram to help show the state transitions

yangdongsheng replied:
Done

dillaman commented on src/librbd/image/RemoveRequest.cc:
```diff
@@ -0,0 +1,641 @@
+    send_close_image(*result);
+    return nullptr;
+  }
+  if (m_watchers.size() > 1) {
```
Needs to merge in new ```filter_out_mirror_watchers``` logic

yangdongsheng replied:
Done

dillaman commented on src/librbd/image/RemoveRequest.cc:
```diff
@@ -0,0 +1,641 @@
+  ldout(m_cct, 20) << ": r=" << *result << dendl;
+
+  if (*result < 0) {
+    if (*result != -ENOENT) {
```
Nit: just test for ```*result == -ENOENT``` first and set the *result to zero if it hits

yangdongsheng replied:
But then we can't print the error message on (*result < 0 && *result != -ENOENT).

dillaman replied:
In this case, -ENOENT isn't an error, though -- we just want to reset the error code in this case

yangdongsheng replied:
Yes, -ENOENT is not an error, so we need to reset the error code for it. But what about the other ones? I mean, if we change it as below:
```
if (*result == -ENOENT)
      *result = 0;
```
Yes, that work. But the error message about other real error would be disappear. please take a look at the whole code.
```
  if (*result < 0) {
    if (*result != -ENOENT) {
      lderr(m_cct) << "error removing child from children list" << cpp_strerror(*result)
                   << dendl;
    } else {
      *result = 0;
    }
  }
```

The difference is the error message gone.

dillaman replied:
@yangdongsheng Why would the error message disappear? The code should be:

```
if (*result == -ENOENT) {
  *result = 0;
} else if (*result < 0) {
  lderr(m_cct) ....
}
```

yangdongsheng replied:
okey, this works.

dillaman commented:
@yangdongsheng This needs a ```test_mock_RemoveRequest``` unit test case to exercise the paths since it is a pretty complex state machine.

yangdongsheng commented:
@trociny and @dillaman Most of the comments are already addressed, but the test_remove_request is still in cooking.

trociny commented:
@yangdongsheng After rebase you need to add this recently merged PR #12549 to asynchronous image removal state machine.

trociny commented:
@yangdongsheng It looks like git modules are incorrectly updated in "asynchronous image removal state machine" commit.

yangdongsheng commented:
Mykola and Jason, Sorry I am in a business trip now, will update my 
patches next week. thanx

On 01/12/2017 03:10 PM, Mykola Golub wrote:
>
> @yangdongsheng <https://github.com/yangdongsheng> It looks like git 
> modules are incorrectly updated in "asynchronous image removal state 
> machine" commit.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/ceph/ceph/pull/12102#issuecomment-272092748>, or 
> mute the thread 
> <https://github.com/notifications/unsubscribe-auth/AC35PVWWDUf9tRnzccvrIS29UGMwKDBQks5rRdH0gaJpZM4K4Plj>.
>

dillaman commented:
@yangdongsheng The changes from PR #12885 will need to be merged into this PR

yangdongsheng commented:
Okey, will rebase against master. BTW, the unittest for removerequest is 
almost ready, will be pushed togather soon.

On 01/17/2017 09:48 PM, Jason Dillaman wrote:
>
> @yangdongsheng <https://github.com/yangdongsheng> The changes from PR 
> #12885 <https://github.com/ceph/ceph/pull/12885> will need to be 
> merged into this PR
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/ceph/ceph/pull/12102#issuecomment-273163475>, or 
> mute the thread 
> <https://github.com/notifications/unsubscribe-auth/AC35PUyWc_3vl24r72SJWKDYNzwv3NSPks5rTMaigaJpZM4K4Plj>.
>

dillaman commented:
@yangdongsheng Nit: can you update your commit prefixes ("cls / rbd" -> "cls_rbd" or "cls/rbd"; "librbd / mirror" -> "librbd"

dillaman commented on src/librbd/image/RemoveRequest.cc:
```diff
@@ -0,0 +1,668 @@
+  if (*result < 0) {
+    lderr(m_cct) << "error opening image: " << cpp_strerror(*result) << dendl;
+    m_retval = *result;
+    mirror_image_remove();
```
Why does it invoke this state only on error? Looking at the current removal code, this state should be between removing the object map and removing the id object.

dillaman commented on src/librbd/internal.cc:
```diff
@@ -164,46 +164,6 @@ int mirror_image_disable_internal(ImageCtx *ictx, bool force,
   return 0;
 }
 
-void filter_out_mirror_watchers(ImageCtx *ictx,
```
Nit: removed this one commit too early since it's still in use in this file.

dillaman commented on src/librbd/image/RemoveRequest.h:
```diff
@@ -27,6 +27,61 @@ class RemoveRequest {
   void send();
 
 private:
+  /**
```
Nit: can you squash this down into 0897d21

dillaman commented on src/librbd/image/RemoveRequest.h:
```diff
@@ -27,6 +27,61 @@ class RemoveRequest {
+   * |               |---<-----------/   v            |         |
+   * |               |             DISABLE FEATURES   (V1)      |
+   * |               |                /  |            |         v
+   * |               |------<--------/   v            | REMOVE MIRROR IMAGE
```
Not sure why REMOVE_MIRROR_IMAGE is only on the error path

dillaman commented on src/librbd/image/RemoveRequest.h:
```diff
@@ -27,6 +27,61 @@ class RemoveRequest {
+   * |               |            VALIDATE IMAGE REMOVAL<-      |
+   * |               |                   |                      v
+   * |               |                   v                      |
+   * |               |               CHECK SNAPS                |
```
Nit: the "check XYZ" actions are not really async states and can probably be left off the chart

dillaman commented on src/librbd/image/RemoveRequest.h:
```diff
@@ -27,6 +27,61 @@ class RemoveRequest {
+   *                                OPEN IMAGE-------------------
+   *                                     |                      | 
+   *                                     v                      |
+   * (error: bottom up)           CHECK EXCLUSIVE LOCK----      |
```
Nit: not sure what "bottom up" refers to here.

dillaman commented on src/librbd/image/RemoveRequest.cc:
```diff
@@ -58,6 +58,8 @@ Context *RemoveRequest<I>::handle_open_image(int *result) {
 
   if (*result < 0) {
     lderr(m_cct) << "error opening image: " << cpp_strerror(*result) << dendl;
+    if (*result != -ENOENT)
```
Squash this down into the original commit. Still need to switch contexts I believe (and delete the m_image_ctx)

dillaman replied:
Nit: braces around "if" clause

dillaman commented on src/librbd/image/RemoveRequest.h:
```diff
@@ -15,11 +15,11 @@ class ProgressContext;
 namespace image {
 
-template<typename ImageCtxT = ImageCtx>
+template<typename ImageCtxT = ImageCtx, typename ContextWQT = ContextWQ>
```
This shouldn't be required -- you can derive the ContextWQ type if required using traits (see Journal traits for an example).

dillaman commented on src/librbd/image/RemoveRequest.cc:
```diff
@@ -31,7 +31,7 @@ RemoveRequest<I, C>::RemoveRequest(IoCtx &ioctx, const std::string &image_name,
 
-  m_image_ctx = new I((m_image_id.empty() ? m_image_name : std::string()),
+  m_image_ctx = I::create((m_image_id.empty() ? m_image_name : std::string()),
                       m_image_id, nullptr, m_ioctx, false);
```
Nit: spacing

dillaman commented on src/test/librbd/mock/MockContextWQ.h:
```diff
@@ -9,6 +9,7 @@
 namespace librbd {
 
 struct MockContextWQ {
+public:
```
Not required -- structs have default public visibility

dillaman commented on src/test/librbd/mock/MockImageCtx.h:
```diff
@@ -36,6 +36,10 @@ struct MockImageCtx {
     return s_instance;
   }
+  void destroy() {
+    return;
```
Switch to ```MOCK_METHOD0(destroy, void())```

dillaman replied:
Still applies -- replace this definition w/ the ```MOCK_METHOD0``` macro to create a stub mock ```destroy``` method

dillaman commented on src/test/librbd/image/test_mock_RemoveRequest.cc:
```diff
@@ -0,0 +1,315 @@
+  librbd::NoOpProgressContext no_op;
+  MockContextWQ op_work_queue;
+  MockTrimRequest mock_trim_request;
+
```
Recommend adding "::testing::InSequence seq" before your first expect statement in each test -- it keeps the test cleaner and ensures the state machine is firing in the correct order

dillaman commented on src/test/librbd/image/test_mock_RemoveRequest.cc:
```diff
@@ -0,0 +1,315 @@
+  expect_trim(*m_mock_imctx, mock_trim_request, 0);
+  expect_state_open(*m_mock_imctx, 0);
+  expect_state_close(*m_mock_imctx);
+  expect_op_work_queue(*m_mock_imctx);
```
Currently, this isn't providing much value. You might just be better off keeping the direct instantiation of ContextWQ in RemoveRequest and avoid the extra complexities of mocking it.

yangdongsheng replied:
@dillaman I have addressed all your comments except this one. Could you provide more information here? thanx :)

dillaman commented on src/test/librbd/image/test_mock_RemoveRequest.cc:
```diff
@@ -0,0 +1,315 @@
+// template definitions
+#include "librbd/image/RemoveRequest.cc"
+template class librbd::image::RemoveRequest<librbd::MockImageCtx, librbd::MockContextWQ>;
+#include "librbd/operation/TrimRequest.cc"
```
Shouldn't need to include the definition to the TrimRequest template -- use template specialization to eliminate it.

yangdongsheng commented:
@dillaman I have addressed most of the comments from you, but sorry I have to be AFK about 10 days for Chinese Sprint Festival. So sorry for the late update on this PR.

yangdongsheng commented:
@dillaman Could you help to take a look?

dillaman commented on src/librbd/ObjectMap.cc:
```diff
@@ -49,6 +49,12 @@ int ObjectMap<I>::remove(librados::IoCtx &io_ctx, const std::string &image_id) {
 }
```
Nit: if the ```ObjectMap::remove``` function is no longer used, please remove it

dillaman commented on src/librbd/operation/DisableFeaturesRequest.cc:
```diff
@@ -31,9 +31,10 @@ template <typename I>
+                                                  uint64_t features,
+                                                  bool force)
+  : Request<I>(image_ctx, on_finish, journal_op_tid), m_features(features),
+    m_force(force) { }
```
Nit: closing brace down to next line

dillaman commented on src/librbd/image/RemoveRequest.cc:
```diff
@@ -0,0 +1,764 @@
+  m_image_ctx->destroy();
+  m_image_ctx = nullptr;
+
+  if (m_retval < 0 && m_retval != ENOENT && !m_unknown_format) {
```
```m_retval != -ENOENT```

dillaman replied:
Not sure about that ```!m_unknown_format``` conditional -- if you fail to open the image due to -EPERM, you should fail out here but since ```m_unknown_format``` would be true in that case, you will continue to remove the image.

yangdongsheng replied:
-EPERM will return earlier in handle_open_image();  but yes, there is no need for !m_unknown_format check here.

dillaman commented on src/librbd/image/RemoveRequest.cc:
```diff
@@ -0,0 +1,764 @@
+    return nullptr;
+  }
+
+  if (!m_old_format) {
```
Nit: this check could actually be moved to ```remove_child``` since by definition a v1 image cannot be a parent/clone

dillaman commented on src/librbd/image/RemoveRequest.h:
```diff
@@ -0,0 +1,184 @@
+  bool m_unknown_format = true;
+  ImageCtxT *m_image_ctx;
+
+  int m_retval = 0;
```
Nit: rename to ```m_ret_val```

dillaman commented on src/librbd/image/RemoveRequest.h:
```diff
@@ -0,0 +1,184 @@
+  void send_journal_remove();
+  Context* handle_journal_remove(int *result);
+
+  void send_objectmap_remove();
```
Nit: these two functions should be ```object_map``` instead of ```objectmap```

dillaman commented on src/librbd/image/RemoveRequest.cc:
```diff
@@ -0,0 +1,772 @@
+  if (*result < 0) {
+    if (*result != -ENOENT) {
+      lderr(m_cct) << "error opening image: " << cpp_strerror(*result) << dendl;
+      return m_on_finish;
```
This would leak the ```m_image_ctx``` -- it needs to be deleted

yangdongsheng commented:
Yes, am fixing it.

thanx
On 14/02/2017, 22:10, Jason Dillaman wrote:
>
> *@dillaman* commented on this pull request.
>
> ------------------------------------------------------------------------
>
> In src/librbd/image/RemoveRequest.cc 
> <https://github.com/ceph/ceph/pull/12102#pullrequestreview-21759846>:
>
> > +  ldout(m_cct, 20) << dendl;
> +
> +  using klass = RemoveRequest<I>;
> +  Context *ctx = create_context_callback<klass, &klass::handle_open_image>(this);
> +
> +  m_image_ctx->state->open(true, ctx);
> +}
> +
> +template<typename I>
> +Context *RemoveRequest<I>::handle_open_image(int *result) {
> +  ldout(m_cct, 20) << ": r=" << *result << dendl;
> +
> +  if (*result < 0) {
> +    if (*result != -ENOENT) {
> +      lderr(m_cct) << "error opening image: " << cpp_strerror(*result) << dendl;
> +      return m_on_finish;
>
> This would leak the |m_image_ctx| -- it needs to be deleted
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/ceph/ceph/pull/12102#pullrequestreview-21759846>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe-auth/AC35PRh2jBxIxB8EgN_X91X0Ljt_9D24ks5rcbXwgaJpZM4K4Plj>.
>

yangdongsheng commented:
retest this please

dillaman commented:
@yangdongsheng This passed the majority of the functional tests but failed the OpenStack tempest test cases twice on the same test when trying to remove an image [1][2]

[1] http://qa-proxy.ceph.com/teuthology/jdillaman-2017-02-15_20:14:46-rbd-wip-jd-testing-distro-basic-smithi/818848/remote/smithi035/qemu/client.0/test.log
[2] http://qa-proxy.ceph.com/teuthology/jdillaman-2017-02-15_13:56:37-rbd-wip-jd-testing-distro-basic-smithi/817887/remote/smithi013/qemu/client.0/test.log

yangdongsheng commented:
Okey, will investigate it ASAP.

On 02/16/2017 11:21 AM, Jason Dillaman wrote:
>
> @yangdongsheng <https://github.com/yangdongsheng> This passed the 
> majority of the functional tests but failed the OpenStack tempest test 
> cases twice on the same test when trying to remove an image [1][2]
>
> [1] 
> http://qa-proxy.ceph.com/teuthology/jdillaman-2017-02-15_20:14:46-rbd-wip-jd-testing-distro-basic-smithi/818848/remote/smithi035/qemu/client.0/test.log
> [2] 
> http://qa-proxy.ceph.com/teuthology/jdillaman-2017-02-15_13:56:37-rbd-wip-jd-testing-distro-basic-smithi/817887/remote/smithi013/qemu/client.0/test.log
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/ceph/ceph/pull/12102#issuecomment-280219814>, or 
> mute the thread 
> <https://github.com/notifications/unsubscribe-auth/AC35PYfPtnXIF89rKepmcTR3qFHC_f8Aks5rc8CdgaJpZM4K4Plj>.
>

yangdongsheng commented:
@dillaman From the code in tempest and log, it's about deleting snapshot. But, sorry I don't have resource for a tempest to reproduce it. Can I access the function testing env? Then I can do a pdb in tempest for more information. I am not sure how to reproduce this problem now.

dillaman commented:
@yangdongsheng It actually might just be an existing issue with the OpenStack tempest test case in how it handles cleanup [1].

[1] https://bugs.launchpad.net/cinder/+bug/1648885

yangdongsheng commented:
@dillaman sounds yes. Thanx for your information. :) Then I think we can continue what we need on this PR.

dillaman commented:
retest this please

dillaman commented:
@yangdongsheng there is a unit test failure with the latest re-build request

yangdongsheng commented:
@dillaman sounds like a race condition here. 
```

send_trim_image()
   - RWLock::RLocker owner_lock(m_image_ctx->owner_lock)       |
   - TrimRequest request->send();                                                     |              m_on_finish->complete()
   - ~ owner_locker automatically in call stack being destroyed.     |              send_close_image()
                                                                                                              |              ceph_assert(!m_image_ctx->owner_lock)
                                                                                                                            
```
So if the gc in c++ for the instance of owner_lock is after send_close_image(), this ceph_assert() could fail.

Sounds possible?

dillaman commented:
@yangdongsheng You should never assert on a lock not being held because another thread could always be using the lock in the background (in another thread). In the case of ```TrimRequest```, if the image is zero bytes, I'd imaging that it would immediately exit while still holding the lock. You can wrap the callback provided to ```TrimRequest``` with ```librbd::util::create_async_context_callback``` to ensure that the "handle" method is invoked in a lock-free context.

yangdongsheng commented:
@dillaman thanx jason. updated as below:
```
diff --git a/src/librbd/image/RemoveRequest.cc b/src/librbd/image/RemoveRequest.cc
index 11633c5..3898ab3 100644
--- a/src/librbd/image/RemoveRequest.cc
+++ b/src/librbd/image/RemoveRequest.cc
@@ -24,6 +24,7 @@ namespace image {

 using librados::IoCtx;
 using util::create_context_callback;
+using util::create_async_context_callback;
 using util::create_rados_ack_callback;

 static const std::string IMAGE_CLIENT_ID("");
@@ -288,7 +289,8 @@ void RemoveRequest<I>::trim_image() {
   ldout(m_cct, 20) << dendl;

   using klass = RemoveRequest<I>;
-  Context *ctx = create_context_callback<klass, &klass::handle_trim_image>(this);
+  Context *ctx = create_async_context_callback(
+    *m_image_ctx, create_context_callback<klass, &klass::handle_trim_image>(this));

   RWLock::RLocker owner_lock(m_image_ctx->owner_lock);
   librbd::operation::TrimRequest<I> *req = librbd::operation::TrimRequest<I>::create(
diff --git a/src/test/librbd/image/test_mock_RemoveRequest.cc b/src/test/librbd/image/test_mock_RemoveRequest.cc
index 6b202a4..74c66c1 100644
--- a/src/test/librbd/image/test_mock_RemoveRequest.cc
+++ b/src/test/librbd/image/test_mock_RemoveRequest.cc
@@ -240,9 +240,9 @@ TEST_F(TestMockImageRemoveRequest, SuccessV1) {

   InSequence seq;
   expect_state_open(*m_mock_imctx, 0);
-  expect_op_work_queue(*m_mock_imctx);
   expect_get_group(*m_mock_imctx, 0);
   expect_trim(*m_mock_imctx, mock_trim_request, 0);
+  expect_op_work_queue(*m_mock_imctx);
   expect_state_close(*m_mock_imctx);
   expect_wq_queue(op_work_queue, 0);

@@ -290,10 +290,10 @@ TEST_F(TestMockImageRemoveRequest, SuccessV2) {

   InSequence seq;
   expect_state_open(*m_mock_imctx, 0);
-  expect_op_work_queue(*m_mock_imctx);
   expect_mirror_image_get(*m_mock_imctx, 0);
   expect_get_group(*m_mock_imctx, 0);
   expect_trim(*m_mock_imctx, mock_trim_request, 0);
+  expect_op_work_queue(*m_mock_imctx);
   expect_remove_child(*m_mock_imctx, 0);
   expect_mirror_disable(*m_mock_imctx, mock_mirror_disable_request, 0);
   expect_state_close(*m_mock_imctx);
@@ -324,10 +324,10 @@ TEST_F(TestMockImageRemoveRequest, NotExistsV2) {

   InSequence seq;
   expect_state_open(*m_mock_imctx, 0);
-  expect_op_work_queue(*m_mock_imctx);
   expect_mirror_image_get(*m_mock_imctx, 0);
   expect_get_group(*m_mock_imctx, 0);
   expect_trim(*m_mock_imctx, mock_trim_request, 0);
+  expect_op_work_queue(*m_mock_imctx);
   expect_remove_child(*m_mock_imctx, 0);
   expect_mirror_disable(*m_mock_imctx, mock_mirror_disable_request, 0);
   expect_state_close(*m_mock_imctx);

```

dillaman commented:
@yangdongsheng Looks like the negative assert is still present and I hit it when I run the unit tests locally:

> 8: [ RUN      ] TestMockImageRemoveRequest.SuccessV1
> 8: /home/jdillaman/ceph/src/librbd/image/RemoveRequest.cc: In function 'void librbd::image::RemoveRequest<ImageCtxT>::send_close_image(int) [with ImageCtxT = librbd::MockImageCtx]' thread 7fa3b8b17700 time 2017-02-20 18:43:24.537008
> 8: /home/jdillaman/ceph/src/librbd/image/RemoveRequest.cc: 390: FAILED assert(!m_image_ctx->owner_lock.is_locked())
> 8:  ceph version 12.0.0-619-g5389cae (5389caea91f9881085571952f7b71dda3a0255f7)
> 8:  1: (ceph::__ceph_assert_fail(char const*, char const*, int, char const*)+0x102) [0x7fa3c0902e82]
> 8:  2: (librbd::image::RemoveRequest<librbd::MockImageCtx>::send_close_image(int)+0x17f) [0x555d4b0b309f]
> 8:  3: (librbd::image::RemoveRequest<librbd::MockImageCtx>::handle_trim_image(int*)+0xa3) [0x555d4b0b45f3]
> 8:  4: (librbd::util::detail::C_StateCallbackAdapter<librbd::image::RemoveRequest<librbd::MockImageCtx>, &librbd::image::RemoveRequest<librbd::MockImageCtx>::handle_trim_image, true>::complete(int)+0x1b) [0x555d4b0b708b]
> 8:  5: (ContextWQ::process(Context*)+0xe1) [0x555d4afdb561]
> 8:  6: (ThreadPool::worker(ThreadPool::WorkThread*)+0x1025) [0x7fa3c090c275]
> 8:  7: (ThreadPool::WorkThread::entry()+0x10) [0x7fa3c090d490]
> 8:  8: (()+0x75ca) [0x7fa3c93955ca]
> 8:  9: (clone()+0x6d) [0x7fa3be08c0ed]

yangdongsheng commented:
@dillaman there is a fix for it. And I tested it by hacking some code in trim to delay the releasing of owner_lock to reproduce the race condition. 

```
--- a/src/librbd/image/RemoveRequest.cc
+++ b/src/librbd/image/RemoveRequest.cc
@@ -302,12 +302,16 @@ template<typename I>
 Context *RemoveRequest<I>::handle_trim_image(int *result) {
   ldout(m_cct, 20) << ": r=" << *result << dendl;
 
+  {
+    // Wait the RLocker released in trim_image()
+    RWLock::WLocker owner_lock(m_image_ctx->owner_lock);
+  }
+
   if (*result < 0) {
     lderr(m_cct) << "warning: failed to remove some object(s): " << cpp_strerror(*result)
                  << dendl;
   }
```

dillaman commented on src/librbd/image/RemoveRequest.cc:
```diff
@@ -0,0 +1,777 @@
+  ldout(m_cct, 20) << ": r=" << *result << dendl;
+
+  {
+    // Wait the RLocker released in trim_image()
```
I am not sure why this is even needed. ```librbd::image::CloseRequest``` will eventually take a write-lock on ```ImageCtx::owner_lock```, so any potential for thread races should be cleared at that point long before the lock is destroyed. Did you ever hit a race where something failed because the trim request process was still holding a read-lock?

yangdongsheng replied:
After a second thought, I think there is no necessary for a assert at https://github.com/ceph/ceph/pull/12102/files#diff-7b8bedbcd26c203c38e02a5bd9dbfa1cR394 . This assert was moved from internal.cc which was a single-threading implementation. As you mentioned, CloseRequest will take the write-lock if they want. So we don't have to make sure the owner_lock is not hold outside. So fix as below:
```
--- a/src/librbd/image/RemoveRequest.cc
+++ b/src/librbd/image/RemoveRequest.cc
@@ -302,11 +302,6 @@ template<typename I>
 Context *RemoveRequest<I>::handle_trim_image(int *result) {
   ldout(m_cct, 20) << ": r=" << *result << dendl;
 
-  {
-    // Wait the RLocker released in trim_image()
-    RWLock::WLocker owner_lock(m_image_ctx->owner_lock);
-  }
-
   if (*result < 0) {
     lderr(m_cct) << "warning: failed to remove some object(s): " << cpp_strerror(*result)
                  << dendl;
@@ -391,8 +386,6 @@ void RemoveRequest<I>::send_close_image(int r) {
   ldout(m_cct, 20) << dendl;
 
   m_ret_val = r;
-  assert(!m_image_ctx->owner_lock.is_locked());
-
   using klass = RemoveRequest<I>;
   Context *ctx = create_context_callback<klass, &klass::handle_send_close_image>(this);

```

And I tested it by ``` for ((i=0;i<100000;i++)); do CEPH_LIB=./lib ./bin/unittest_librbd --gtest_filter='TestMockImageRemoveRequest.*' >> /tmp/output; done; grep FAIL /tmp/output ``` with hacking this code:
```
--- a/src/librbd/image/RemoveRequest.cc
+++ b/src/librbd/image/RemoveRequest.cc
@@ -296,6 +296,11 @@ void RemoveRequest<I>::trim_image() {
   librbd::operation::TrimRequest<I> *req = librbd::operation::TrimRequest<I>::create(
     *m_image_ctx, ctx, m_image_ctx->size, 0, m_prog_ctx);
   req->send();
+  int i = 0;
+  for (i = 0; i < 10000; i++) {
+    lderr(m_cct) << "delay delay delay the lock releasing: "
+                 << dendl;
+  }
 }
```

trociny replied:
@yangdongsheng just a hint: `--gtest_repeat=1000000`

yangdongsheng replied:
@trociny , Oh great!! thanx for your information. :)
